### PR TITLE
errors: make WithData merge maps

### PR DESCRIPTION
**Pull request by kr**
This lets us accumulate data in an error rather than
entirely replacing what was there before.

This is the first step in an alternate approach to #115.

### CODE REVIEW

File: errors/errors.go [hunk 1]
```diff
@@ -126,6 +126,29 @@ func WithData(err error, v interface{}) error {
...
+// Note that if err already has a data item of any other type,
+// it will not be accessible via the returned error value.
+func WithDataKV(err error, keyval ...interface{}) error {
+	newkv := make(map[string]interface{})
+	oldkv, _ := Data(err).(map[string]interface{})
+	for k, v := range oldkv {
+		newkv[k] = v
+	}
+	for i := 0; i < len(keyval); i += 2 {
+		newkv[keyval[i].(string)] = keyval[i+1]
```
Comment (erykwalder) [inline errors/errors.go]: There's a potential panic here.

File: errors/errors.go [hunk 2]
```diff
@@ -126,6 +126,29 @@ func WithData(err error, v interface{}) error {
...
+	}
+	return WithData(err, newkv)
```
Comment (erykwalder) [inline errors/errors.go]: Should `WithData` become private? It feels strange to have one method that expects cooperation and another that doesn't.

File: errors/errors.go (hunk 1 above)
Comment (kr) [inline errors/errors.go]: Yeah, I want it to panic if it's not a string. That's why the doc says "Values kN must be strings".

File: errors/errors.go (hunk 2 above)
Comment (kr) [inline errors/errors.go]: Hm actually I'm on the fence over this. We include a json array of action errors as the top level data item. We could just make all data items be maps, but that would add one more nesting layer to an already annoyingly-nested structure.

What do you think?

File: errors/errors.go (hunk 1 above)
Comment (erykwalder) [inline errors/errors.go]: Also from an odd number of args.

File: errors/errors.go (hunk 2 above)
Comment (erykwalder) [inline errors/errors.go]: I guess I'm just worried about accidentally clobbering. Using WithDataKV after WithData will clobber, as will the other way around.

Comment (erykwalder) [inline errors/errors.go]: But if this is what we've got for now, that's fine.

Comment (kr) [inline errors/errors.go]: Ok, I'll make them all be maps.

Comment (erykwalder) [inline errors/errors.go]: Actually, perhaps we could panic if data is not nil or not a map.

File: errors/errors.go (hunk 1 above)
Comment (kr) [inline errors/errors.go]: Same thing for odd number of args. It should panic.
I added a TODO to check for this in vet.

**Comment by kr** [comment]
PTAL

**Comment by kr** [comment]
PTAL @erykwalder @dominic

**Comment by jbowens** [comment]
lgtm

**Pull request merged**